### ISSUE TITLE
Deno support

### ISSUE THREAD

**Issue by michailVestnik** [lovell/sharp#2583]
Hi!
It would be very nice to make an ES import for Deno and place the documentation and examples in the main repository at https://deno.land

For example: 

` import { sharp } from 'https://deno.land/x/sharp@v0.27.0/mod.ts' `

Thanks!

**Comment by lovell** [comment]
Hi, sharp is a "native" module that interfaces with Node.js through [N-API](https://nodejs.org/dist/latest/docs/api/n-api.html). This API isn't currently available in Deno, although there appears to be (or at least have been) a desire to support it - see https://github.com/denoland/deno/issues/237#issuecomment-396239572 and https://github.com/denoland/deno/issues/241

**Comment by shadowtime2000** [comment]
I know very little about native/systems programming and stuff - but couldn't you create Rust bindings to the C++ API and then have it hook into Deno's unstable native plugin API or have it compile to WASM/WASI? Again, I know very little about this stuff.

**Comment by michailVestnik** [comment]
Perhaps this link will be useful to you https://github.com/emscripten-core/emscripten
Deno has very good wasm support!
You can connect the module in several lines: https://deno.land/manual@v1.8.0/getting_started/webassembly

**Comment by shadowtime2000** [comment]
Sorry if this sounds rude, but it doesn't seem like there is an interest from the sharp maintainers so I guess someone from the community could just use emscripten with a simple c++ project importing sharp and just publish that.

**Comment by michailVestnik** [comment]
> Sorry if this sounds rude, but it doesn't seem like there is an interest from the sharp maintainers so I guess someone from the community could just use emscripten with a simple c++ project importing sharp and just publish that.

At the moment, there are no photo processing modules for Deno.
I was forced to make an interface for the most popular library for photo processing in Python - Pillow.
If you're interested, I can post my code on Github in the coming days.

**Comment by tomByrer** [comment]
Here is the lib mentioned above:
https://github.com/michailVestnik/python_pillow

@michailVestnik can you reopen please?  I & I'm sure others would love to use this lib on Deno someday....

**Comment by jgod** [comment]
Would be killer

**Comment by lovell** [comment]
For those who haven't yet seen, Deno is in the process of adding support for Node-API (N-API) via https://github.com/denoland/deno/pull/13633

Once this is made available we can assess how much additional work, if any, would be required by sharp to integrate.

**Comment by MarkBennett** [comment]
Looks like denoland/deno#13633 was merged yesterday @lovell! 😄

**Comment by deeprobin** [comment]
Why is there no deno-native implementation?

**Comment by lino-levan** [comment]
Node API (formerly NAPI), was implemented into Deno a while ago. Could this issue be opened again, I think it deserves some discussion

**Comment by lovell** [comment]
> Node API (formerly NAPI), was implemented into Deno a while ago

The merged PR was part of Deno [v1.26.1](https://github.com/denoland/deno/releases/tag/v1.26.1) published on the 6th October, 35 days ago. I haven't yet seen any documentation or examples relating to this feature.

Has anyone tried using sharp with Deno since then? Happy to reopen if there are specific things that need to be done in sharp.

I note Deno has its own dependency management, and features of the Node.js/npm/yarn ecosystem such as install scripts are not (yet?) available, so it's unclear how to get platform and architecture specific binaries installed. Perhaps manually-managed import maps might be required to handle this?

**Comment by lovell** [comment]
A quick check reveals that the latest Deno allows me to use sharp, albeit after running `npm install` within the Deno cache and with `napi_add_finalizer is not yet supported` warnings (ignorable?).

```sh
$ deno --version
deno 1.27.2 (release, x86_64-unknown-linux-gnu)
v8 10.8.168.4
typescript 4.8.3
```
Here's my `test.ts` file:
```js
import sharp from 'npm:sharp@0.31.2'
console.log(sharp.format)
```
Running it first time populates Deno's cache from npm, but doesn't install binaries and fails.
```sh
$ cat test.ts | deno run --allow-all --unstable -
error: Uncaught Error: 
Something went wrong installing the "sharp" module
...
```
Manually install the binaries by running `npm install` within Deno's cache for now (perhaps Deno will do this bit for you at some point).
```sh
$ cd ~/.cache/deno/npm/registry.npmjs.org/sharp/0.31.2
$ npm install
```
Running the test script again appears to work:
```sh
$ cat test.ts | deno run --allow-all --unstable -
napi_add_finalizer is not yet supported.
...
{
  jpeg: {
    id: "jpeg",
    input: { file: true, buffer: true, stream: true, fileSuffix: [ ".jpg", ".jpeg", ".jpe" ] },
    output: { file: true, buffer: true, stream: true, alias: [ "jpe", "jpg" ] }
  },
...
```
Is anyone aware of a better/easier way to get Deno to run install scripts, rather than manually running `npm install`?

In summary, we should add something to the documentation about this, but it looks like there are no code changes required. :rocket:

**Comment by deeprobin** [comment]
![grafik](https://user-images.githubusercontent.com/10373452/201166827-5917ecc7-1a8e-40f8-a859-31439d1fe52d.png)

Why was this closed?

/cc @lovell

**Comment by lovell** [comment]
@deeprobin The person who opened the issue also closed it (on the 7th April 2021). Had there been an obvious change required to sharp then I would happily have reopened it.

At the time it appeared that Deno was planning to add support for Node-API, and now they have and it works, which is great news, and I'm sure you'll join me in thanking the Deno maintainers for this.

**Comment by jgod** [comment]
Thanks Lovell. 

I can maybe port some of my old node stuff over to Deno now, which I find more maintainable (less dev tooling/packaging overhead) for little scripts/CLIs.

**Comment by UrielCh** [comment]
I have tried on a windows host

nodeJS version: v18.12.0

deno version: 1.29.4

with `import sharp from 'npm:sharp@0.31.3';`

and `import sharp from 'npm:sharp@0.31.2';`


I start the npm install from the np, cached folded, and I still get the error:

```ts
deno run -A --unstable .\test-sharp.ts
error: Uncaught Error: 
Something went wrong installing the "sharp" module

Error in native callback

Possible solutions:
- Install with verbose logging and look for errors: "npm install --ignore-scripts=false --foreground-scripts --verbose sharp"
- Install for the current win32-x64 runtime: "npm install --platform=win32 --arch=x64 sharp"
- Consult the installation documentation: https://sharp.pixelplumbing.com/install
    at Object.<anonymous> (file:///C:/Users/uriel/AppData/Local/deno/npm/registry.npmjs.org/sharp/0.31.2/lib/sharp.js:34:9)
    at Object.<anonymous> (file:///C:/Users/uriel/AppData/Local/deno/npm/registry.npmjs.org/sharp/0.31.2/lib/sharp.js:37:4)
    at Module._compile (deno:ext/node/02_require.js:747:36)
    at Object.Module._extensions..js (deno:ext/node/02_require.js:780:12)
    at Module.load (deno:ext/node/02_require.js:658:34)
    at Function.Module._load (deno:ext/node/02_require.js:515:14)
    at Module.require (deno:ext/node/02_require.js:680:21)
    at require (deno:ext/node/02_require.js:820:18)
    at Object.<anonymous> (file:///C:/Users/uriel/AppData/Local/deno/npm/registry.npmjs.org/sharp/0.31.2/lib/constructor.js:8:1)
    at Object.<anonymous> (file:///C:/Users/uriel/AppData/Local/deno/npm/registry.npmjs.org/sharp/0.31.2/lib/constructor.js:425:4)
```

I all so tried to run `npm install --ignore-scripts=false --foreground-scripts --verbose sharp` in deno cache npm package

and with `npm install --platform=win32 --arch=x64 sharp`

Same error.

**Comment by xenova** [comment]
Are there any updates on this?

**Comment by lovell** [comment]
The somewhat manual steps outlined in https://github.com/lovell/sharp/issues/2583#issuecomment-1310628079 should already allow the use of sharp with Deno.

Installation will hopefully be much simpler when the feature being tracked at https://github.com/denoland/deno/issues/16164 is made available, at which point we can do things like add Deno to CI environments and update documentation.

**Comment by xenova** [comment]
> The somewhat manual steps outlined in [#2583 (comment)](https://github.com/lovell/sharp/issues/2583#issuecomment-1310628079) should already allow the use of sharp with Deno.
> 
> Installation will hopefully be much simpler when the feature being tracked at [denoland/deno#16164](https://github.com/denoland/deno/issues/16164) is made available, at which point we can do things like add Deno to CI environments and update documentation.

Great, thanks for the quick response.

**Comment by VelinGeorgiev** [comment]
I manually installed Sharp as suggested above, but I am getting the following error on Mac M1. Looks like I will park Deno - Sharp experiments for now.

```
error: Uncaught Error: 
Something went wrong installing the "sharp" module

Unable to find napi_register_module_v1 symbol in /Users/xx/Library/Caches/deno/npm/registry.npmjs.org/sharp/0.32.4/build/Release/sharp-darwin-arm64v8.node

Possible solutions:
- Install with verbose logging and look for errors: "npm install --ignore-scripts=false --foreground-scripts --verbose sharp"
- Install for the current darwin-arm64v8 runtime: "npm install --platform=darwin --arch=arm64v8 sharp"
- Consult the installation documentation: https://sharp.pixelplumbing.com/install
    at Object.<anonymous> (file:///Users/xx/Library/Caches/deno/npm/registry.npmjs.org/sharp/0.32.4/lib/sharp.js:37:9)
    at Object.<anonymous> (file:///Users/xx/Library/Caches/deno/npm/registry.npmjs.org/sharp/0.32.4/lib/sharp.js:40:4)
    at Module._compile (node:module:718:36)
    at Object.Module._extensions..js (node:module:737:12)
    at Module.load (node:module:652:34)
    at Function.Module._load (node:module:534:16)
    at Module.require (node:module:671:23)
    at require (node:module:771:20)
    at Object.<anonymous> (file:///Users/xx/Library/Caches/deno/npm/registry.npmjs.org/sharp/0.32.4/lib/constructor.js:11:1)
    at Object.<anonymous> (file:///Users/xx/Library/Caches/deno/npm/registry.npmjs.org/sharp/0.32.4/lib/constructor.js:441:4)
```

**Comment by lovell** [comment]
@VelinGeorgiev Best guess would be that https://github.com/denoland/deno/commit/512d5337c480a2a2704881d3fe1c40b6e0445cf0 altered Deno to expect a native module to register itself using `NODE_API_ADDON` / `napi_register_module_v1` and therefore drop support for `NODE_API_MODULE` / `napi_module_register`. If intentional, this sounds like a semver major change as I think this will pin the minimum Node-API version to 8. Perhaps ask at the Deno repo about this - https://github.com/nodejs/node/issues/45657 is relevant.

**Comment by VelinGeorgiev** [comment]
Thanks @lovell! I will ask at the Deno repo.

**Comment by oscarotero** [comment]
> Is anyone aware of a better/easier way to get Deno to run install scripts, rather than manually running `npm install`?
> 
> In summary, we should add something to the documentation about this, but it looks like there are no code changes required. 🚀

@lovell I propose to change the way the binary is installed in sharp. Instead of using the `install` script, use dependencies. This allows to use the package in Deno without requiring any manual step.

For example, Esbuild publishes a binary for each platform/architecture supported as `optionalDependencies`: https://github.com/evanw/esbuild/blob/main/npm/esbuild/package.json

Pagefind does something similar: https://github.com/CloudCannon/pagefind/blob/main/wrappers/node/package.json

The package manager (NPM or Deno) pick automatically the dependency [based on this suffix](https://github.com/denoland/deno/issues/15611#issuecomment-1330776743).

**Comment by lovell** [comment]
@some-user123 For `deno compile` please see https://github.com/lovell/sharp/issues/3912
